src/main/java/me/moodcat/api/filters/RateLimitFilter.java

```diff
@@ -0,0 +1,189 @@
+package me.moodcat.api.filters;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.List;
+
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+import me.moodcat.api.models.ExceptionResponse;
+import me.moodcat.database.entities.User;
+
+import org.jboss.resteasy.core.interception.PostMatchContainerRequestContext;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
+import com.google.inject.name.Named;
+
+/**
+ * The {@code RateLimitFilter} rejects filters based on the {@link RateLimit} annotation
+ * on resource methods. It is used to filter out abusive messages.
+ */
+@Slf4j
+@RateLimit
+@Singleton
+@javax.ws.rs.ext.Provider
+public class RateLimitFilter implements ContainerRequestFilter {
+
+    private static final int HTTP_TOO_MANY_REQUESTS = 429;
+
+    /**
+     * A map that temporary stores requests for a given method.
+     */
+    private final Multimap<Method, Entry> entries;
```

Comment:
Also we have a lot of synchronized blocks. There is no concurrent map that can serve the same purpose?